include/illixr/ProperQuaternion.hpp

```diff
@@ -0,0 +1,232 @@
+#pragma once
+#include <eigen3/Eigen/Dense>
+#include <iostream>
+
+namespace ILLIXR {
+
+/**
+ * @brief A more complete implementation of the Eigen::Quaternion
+ *
+ * The base class, Eigen::Quaternion, is missing some arithmetic functionality, like scalar multiplication,
+ * this implementation adds this functionality.
+ *
+ * @tparam Scalar_ The type of data the Quaternion holds (typically double ot float)
+ * @tparam Options_ Data alignment
+ */
+template<typename Scalar_, int Options_ = Eigen::AutoAlign>
+class ProperQuaternion : public Eigen::Quaternion<Scalar_, Options_> {
+public:
+    typedef typename Eigen::Quaternion<Scalar_>::Base Base;
+
+    using Base::operator=;
+
+    /**
+     * @brief Copy operator
+     * @param other A ProperQuaternion instance
+     * @return a copy
+     */
+    inline ProperQuaternion& operator=(const ProperQuaternion& other) {
+        Base::operator=(other);
+        return *this;
+    }
+
+    using Base::operator*=;
+
+    // constructors
+    ProperQuaternion()                              = default;
+    ProperQuaternion(const ProperQuaternion& other) = default;
+
+    /**
+     * Copy constructor from the base class
+     * @param other An Eigen::Quaternion instance
+     */
+    [[maybe_unused]] explicit ProperQuaternion(const Eigen::Quaternion<Scalar_, Options_>& other)
+        : Eigen::Quaternion<Scalar_, Options_>(other) { }
+
+    /**
+     * Constructor from individual components
+     * @param w The w component
+     * @param x The x component
+     * @param y The y component
+     * @param z The z component
+     */
+    [[maybe_unused]] ProperQuaternion(const Scalar_& w, const Scalar_& x, const Scalar_& y, const Scalar_& z)
+        : Eigen::Quaternion<Scalar_, Options_>(w, x, y, z) { }
+
+    /**
+     * Constructor from a scalar and vector
+     * @tparam Derived
+     * @param w The w component
+     * @param vec The x, y, and z components
+     */
+    template<typename Derived>
+    [[maybe_unused]] ProperQuaternion(const Scalar_& w, const Eigen::MatrixBase<Derived>& vec)
+        : Eigen::Quaternion<Scalar_, Options_>(w, vec) { }
+
+    /**
+     * Constructor from a pointer array (assumes 4 elements)
+     * @param data Pointer to the first element
+     */
+    [[maybe_unused]] explicit ProperQuaternion(const Scalar_* data)
+        : Eigen::Quaternion<Scalar_, Options_>(data) { }
+
+    /**
+     * Constructor from a vector. Assumes the vector is x, y, z, w order.
+     * @param vec The input vector
+     */
+    [[maybe_unused]] explicit ProperQuaternion(const Eigen::Vector<Scalar_, 4>& vec)
```

Comment:
Building reports error about the use of `Eigen::Vector`

Reply:
Can you post the errors you are getting and what compiler and version as well? I do not get any errors when I try to compile.

Reply:
There are some of the errors:
```/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp:77:61: error: ‘Vector’ in namespace ‘Eigen’ does not name a template type; did you mean ‘VectorXd’?
   77 |     [[maybe_unused]] explicit ProperQuaternion(const Eigen::Vector<Scalar_, 4>& vec)
      |                                                             ^~~~~~
      |                                                             VectorXd
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp:77:67: error: expected ‘,’ or ‘...’ before ‘<’ token
   77 |     [[maybe_unused]] explicit ProperQuaternion(const Eigen::Vector<Scalar_, 4>& vec)
      |                                                                   ^
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp:142:12: error: ‘Vector4’ in namespace ‘Eigen’ does not name a template type; did you mean ‘Vector4d’?
  142 |     Eigen::Vector4<Scalar_> asVector() const {
      |            ^~~~~~~
      |            Vector4d
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp: In constructor ‘ILLIXR::ProperQuaternion<Scalar_, Options_>::ProperQuaternion(int)’:
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp:78:48: error: ‘vec’ was not declared in this scope
   78 |         : Eigen::Quaternion<Scalar_, Options_>(vec[3], vec[0], vec[1], vec[2]) { }
      |                                                ^~~
In file included from /home/qinjunj/Research/PR3.3/ILLIXR_rk4/plugins/pose_prediction/plugin.cpp:6:
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/runge-kutta.hpp: In function ‘ILLIXR::ProperQuaterniond ILLIXR::q_dot(const Vector3d&, const ProperQuaterniond&)’:
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/runge-kutta.hpp:54:71: error: ‘const ProperQuaterniond’ {aka ‘const class ILLIXR::ProperQuaternion<double>’} has no member named ‘asVector’
   54 |     return ProperQuaterniond(Eigen::Vector4d(0.5 * makeOmega(av) * dq.asVector()));
      |                                                                       ^~~~~~~~
In file included from /home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/runge-kutta.hpp:2,
                 from /home/qinjunj/Research/PR3.3/ILLIXR_rk4/plugins/rk4_integrator/plugin.cpp:10:
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp:77:61: error: ‘Vector’ in namespace ‘Eigen’ does not name a template type; did you mean ‘VectorXd’?
   77 |     [[maybe_unused]] explicit ProperQuaternion(const Eigen::Vector<Scalar_, 4>& vec)
      |                                                             ^~~~~~
      |                                                             VectorXd
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp:77:67: error: expected ‘,’ or ‘...’ before ‘<’ token
   77 |     [[maybe_unused]] explicit ProperQuaternion(const Eigen::Vector<Scalar_, 4>& vec)
      |                                                                   ^
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp:142:12: error: ‘Vector4’ in namespace ‘Eigen’ does not name a template type; did you mean ‘Vector4d’?
  142 |     Eigen::Vector4<Scalar_> asVector() const {
      |            ^~~~~~~
      |            Vector4d
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp: In constructor ‘ILLIXR::ProperQuaternion<Scalar_, Options_>::ProperQuaternion(int)’:
/home/qinjunj/Research/PR3.3/ILLIXR_rk4/include/illixr/ProperQuaternion.hpp:78:48: error: ‘vec’ was not declared in this scope
   78 |         : Eigen::Quaternion<Scalar_, Options_>(vec[3], vec[0], vec[1], vec[2]) { }
      |                                                ^~~
```

Reply:
what version of Eigen do you have installed?

Reply:
3.3.7-2

Reply:
The issue is with the use of Eigen::Vector or Eigen::Vector4 as a template. Are they defined somewhere?

Reply:
Changing `Eigen::Vector<Scalar_, 4>` and `Eigen::Vector4<Scalar_>` to `Eigen::Vector4d` would fix the compiling issue

Reply:
They are defines inside of a series of burried macros in Eigen/Core/Matrix.h
The version you are using is 6+ years old, and appears to be missing these definitions.

We can do that, but we should be sure it does not break anything that expects a float matrix. 

I can work on making these changes (or you can if you have time), but it will have to wait for the hand tracking stuff to be finished.

Reply:
We should stick with the typedefs, which is also what other plugins like [timewarp](https://github.com/ILLIXR/ILLIXR/blob/20d19579ab1b16638bbac6045cbd8c88504b1c93/plugins/timewarp_gl/plugin.cpp#L280) are doing. 

I can try getting to these changes.